docs/manual/src/foreign_traits.md

```diff
@@ -0,0 +1,144 @@
+}
+```
+
+If you are using macros add `#[uniffi::export]` above the trait.
+Otherwise define this trait in your UDL file:
+
+```webidl
+[Trait]
+interface Keychain {
+    [Throws=KeyChainError]
+    string? get(string key);
+
+    [Throws=KeyChainError]
+    void put(string key, string data);
+};
+```
+
+## 2. Allow it to be passed into Rust
+
+Here, we define a new object with a constructor which takes a keychain.
+
+```webidl
+interface Authenticator {
+    constructor(Keychain keychain);
+    void login();
+};
+```
+
+In Rust we'd write:
+
+```rust,no_run
+struct Authenticator {
+  keychain: Arc<dyn Keychain>,
+}
+
+impl Authenticator {
+  pub fn new(keychain: Arc<dyn Keychain>) -> Self {
+    Self { keychain }
+  }
+
+  pub fn login(&self) {
+    let username = self.keychain.get("username".into());
+    let password = self.keychain.get("password".into());
+  }
+}
+```
+
+## 3. Create a foreign language implementation of the trait
+
+Here's a Kotlin implementation:
+
+```kotlin
+class KotlinKeychain: Keychain {
+    override fun get(key: String): String? {
+        // … elide the implementation.
+        return value
+    }
+    override fun put(key: String) {
+        // … elide the implementation.
+    }
+}
+```
+
+…and Swift:
+
+```swift
+class SwiftKeychain: Keychain {
+    func get(key: String) -> String? {
+        // … elide the implementation.
+        return value
+    }
+    func put(key: String) {
+        // … elide the implementation.
+    }
+}
+```
+
+## 4. Pass the implementation to Rust
+
+Again, in Kotlin
+
+```kt
+val authenticator = Authenticator(KotlinKeychain())
+// later on:
+authenticator.login()
+```
+
+and in Swift:
+
+```swift
+let authenticator = Authenticator(SwiftKeychain())
+// later on:
+authenticator.login()
+```
+
+Care is taken to ensure that things are cleaned up in the foreign language once all Rust references drop.
+
+## ⚠️  Avoid cycles
+
+Foreign trait implementations make it easy to create cycles between Rust and foreign objects causing memory leaks.
+For example a foreign implementation holding a reference to a Rust object which also holds a reference to the same foreign implementation.
+
+UniFFI doesn't try to help here and there's no universal advice; take the usual precautions.
+
+# Error handling
+
+We must handle foreign code failing, so all methods of the Rust trait should return a `Result<>` with a [compatible error type](./udl/errors.md) otherwise these errors will panic.
+
+## Unexpected Error handling.
+
+So long as your function returns a `Result<>`, it's possible for you to define how "unexpected" errors
+(ie, errors not directly covered by your `Result<>` type, panics, etc) are converted to your `Result<>`'s `Err`.
+
+If your code defines a `From<uniffi::UnexpectedUniFFICallbackError>` impl for your error type, then those errors will be converted into your error type which will be returned to the Rust caller instead of panicing.
```

Comment:
This doc looks great, but maybe call out the panic slightly more.  Something like: "...your error type which will be returned to the Rust caller.  If they do not, then the generated code  will panic."